examples/NanoGPT/main.jl

```diff
@@ -0,0 +1,259 @@
+        output_length::Int=1024
+)
+    rng = Random.default_rng()
+    Random.seed!(rng, 1234)
+
+    dev = reactant_device()
+    cdev = cpu_device()
+
+    if inference
+        @printf "[Info] Inference mode enabled.\n"
+
+        @assert !isempty(model_path) "Please provide a path to a model checkpoint."
+
+        @printf "[Info] Loading model from %s.\n" model_path
+        model_config = JLD2.load(model_path, "model_config")
+        model = GPT(; model_config...)
+        ps = JLD2.load(model_path, "parameters")
+        st = JLD2.load(model_path, "states")
+        alphabet = JLD2.load(model_path, "alphabet")
+        sequence_length = model_config.sequence_length
+
+        texts = generate_text(
+            model, ps, st, seed; alphabet, output_length, sequence_length
+        )
+
+        for (i, (text, s)) in enumerate(zip(texts, seed))
+            @printf "[Info] Seed [%d]: %s\n" i s
+            @printf "[Generated Text] %s\n\n" text
+        end
+
+        return
+    end
+
+    alphabet, trainX, trainY, testX, testY = get_nanogpt_data(; sequence_length, test_split)
+
+    @printf "[Info] Alphabet size: %d\n" length(alphabet)
+    @printf "[Info] Training size: %d sequences.\n" size(trainX, 2)
+    @printf "[Info] Testing  size: %d sequences.\n\n" size(testX, 2)
+
+    train_loader = DataLoader(
+        (trainX, trainY); batchsize, shuffle=true, parallel=true
+    ) |> dev
+
+    model_config = (;
+        n_vocab=length(alphabet), n_embed, sequence_length, n_hidden,
+        n_layers, dropout_rate, n_heads, qk_dim, v_dim
+    )
+    model = GPT(; model_config...)
+    ps, st = Lux.setup(rng, model) |> dev
+    @printf "[Info] Number of parameters: %d\n" Lux.parameterlength(ps)
+    @printf "[Info] Number of states: %d\n\n" Lux.statelength(st)
+
+    opt = Adam(lr)
+    train_state = Training.TrainState(model, ps, st, opt)
+
+    @printf "[Info] Compiling Inference Model...\n"
+    testX, testY = (testX, testY) |> dev
+    start_time = time()
+    model_compiled = @compile model(testX, ps, Lux.testmode(st))
+    time_to_compile = time() - start_time
+    best_test_loss = Inf
+
+    @printf "[Info] Time taken to compile inference model: %0.5fs\n" time_to_compile
+    @printf "[Info] Starting Model Training...\n\n"
+
+    loss_fn = CrossEntropyLoss(; logits=Val(true))
+
+    iter = 0
+    for epoch in 1:epochs
+        for (x, y) in train_loader
+            iter += 1
+
+            start_time = time()
+            _, loss, _, train_state = Training.single_train_step!(
+                AutoEnzyme(), loss_fn, (x, y), train_state
+            )
+            time_taken = time() - start_time
+
+            if iter % 100 == 0
+                @printf "[Train] Epoch %3d\tIteration %6d\tLoss %.8e\tTime per \
+                         Iteration %0.5f\n" epoch iter loss time_taken
+            end
+        end
+
+        test_loss = loss_fn(
+            Array(first(model_compiled(testX, ps, Lux.testmode(st)))), testY
```

Comment:
```suggestion
            first(model_compiled(testX, ps, Lux.testmode(st))), testY
```

Reply:
We need to compile the loss_fn for the non-`Array` version to work

Reply:
Ah yeah this might not be needed for your case here